File: Difficulty: Hard/Longest valid Parentheses/longest-valid-parentheses.cpp
//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function Template for C++

class Solution {
  public:
    int maxLength(string str) {
        // Get the length of the input string
        int n = str.length();
        
        // Stack to store indices of parentheses
        stack<int> st;
        
        // Initial push of -1 to handle edge cases (e.g., valid string starting from index 0)
        st.push(-1);
        
        // Variable to store the maximum valid length
        int ans = 0;
        
        // Traverse the input string character by character
        for(int i = 0; i < n; ++i) {
            // If the current character is '(', push its index onto the stack
            if(str[i] == '(') {
                st.push(i);
            }
            // If the current character is ')', attempt to close the most recent '('
            else {
                st.pop(); // Pop the top element (an opening parenthesis index)
                
                // If the stack becomes empty, it means we have unmatched closing parenthesis
                if(st.empty()) {
                    // Push the index of the unmatched closing parenthesis as a new "starting point"
                    st.push(i);
                }
                // Otherwise, calculate the length of the valid substring
                else {
                    // Calculate the length between current index and the top of the stack
                    ans = max(ans, i - st.top());
                }
            }
        }
        
        // Return the maximum length of valid parentheses found
        return ans;
    }
};


//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        string S;
        cin >> S;

        Solution ob;
        cout << ob.maxLength(S) << "\n";
    }
    return 0;
}
// } Driver Code Ends

File: Difficulty: Hard/Longest valid Parentheses/longest-valid-parentheses.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

class GFG {
    public static void main(String args[]) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(in.readLine());
        while (t-- > 0) {
            String S = in.readLine();

            Solution ob = new Solution();
            System.out.println(ob.maxLength(S));

            System.out.println("~");
        }
    }
}
// } Driver Code Ends



class Solution {
    static int maxLength(String s) {
       int n=s.length();
       int l=0;
       int r=0;
       int max=0;
       for(int  i=0;i<n;i++)
       {
           if(s.charAt(i)=='(')
           l++;
           if(s.charAt(i)==')')
           r++;
           if(l==r)
           max=Math.max(max,2*l);
           if(r>l)
           {
               l=0;
               r=0;
           }
       }
       r=0;
       l=0;
       for(int i=n-1;i>=0;i--)
       {
           if(s.charAt(i)=='(')
           l++;
           if(s.charAt(i)==')')
           r++;
           if(l==r)
           max=Math.max(max,2*l);
           if(r<l)
           {
               l=0;
               r=0;
           }
       }
       return max;
    }
}